File: Algorithm-Repository/Sorting-Algorithms/Cpp/BubbleSort.cpp
// BubbleSort.cpp
// Language: C++
// Topic: Sorting Algorithms
// Purpose: Sort an array of integers in ascending order using Bubble Sort.
// Time Complexity: O(n^2) in worst & average case, O(n) in best case (already sorted)
// Space Complexity: O(1) — In-place sorting (no extra memory required)


#include <iostream>
using namespace std;

/**
 * 🧠 Topic Explanation — Bubble Sort
 * ------------------------------------
 * Bubble Sort is a simple comparison-based sorting algorithm.
 * The algorithm repeatedly steps through the array, compares adjacent elements,
 * and swaps them if they are in the wrong order (e.g., arr[j] > arr[j+1] for ascending order).
 * 
 * This process continues until no more swaps are needed, meaning the array is sorted.
 * After each pass:
 *  - The largest unsorted element “bubbles up” to its correct position at the end.
 *  - The next pass can ignore the last sorted elements.
 * 
 * ✅ Stable Algorithm — does not change the relative order of equal elements.
 * ✅ In-place Algorithm — requires no extra space beyond the input array.
 * ⚠️ Not efficient for large datasets — better algorithms like Merge Sort or Quick Sort are preferred.
 *
 * 📊 Time Complexity:
 * -------------------
 * - Best Case (Already Sorted):       O(n)       → thanks to the early exit check.
 * - Average Case (Random Order):      O(n^2)
 * - Worst Case (Reverse Order):       O(n^2)
 *
 * 🧮 Space Complexity:
 * -------------------
 * - O(1) — only a few extra variables, sorting is done in-place.
 *
 * 📌 Use Cases:
 * ------------
 * - Teaching/learning basic sorting concepts.
 * - Sorting very small arrays.
 * - Useful when array is nearly sorted (early termination gives O(n) time).
 */
void bubbleSort(int arr[], int n)
{

    bool swap;
    for (int i = 0; i < n - 1; i++)
    {
        swap = false;
        for (int j = 0; j < n - i - 1; j++)
        {
            if (arr[j] > arr[j + 1])
            {
                swap = true;
                int temp = arr[j];
                arr[j] = arr[j + 1];
                arr[j + 1] = temp;
            }
        }
        if (!swap) break;
    }
}

/**
 * Utility function to print an array
 */
void printArray(int arr[], int n) {
    for (int i = 0; i < n; i++)
        cout << arr[i] << " ";
    cout << endl;
}

/**
 * 🧪 Sample Input/Output:
 * -----------------------
 * Input:
 *  Array before sorting: 30 25 20 15 10
 * 
 * Output:
 *  Array after sorting:  10 15 20 25 30
 */

int main()
{
    int arr[] = {30, 25, 20, 15, 10};
    int n = sizeof(arr) / sizeof(arr[0]);
    
    cout << "Array before sorting: ";
    printArray(arr, n);

    bubbleSort(arr, n);

    cout << "Array after sorting: ";
    printArray(arr, n);

    return 0;
}

File: Algorithm-Repository/Sorting-Algorithms/Java/BubbleSort.java
// BubbleSort.java
// Language: Java
// ✅ Short Description:
// Bubble Sort is a simple comparison-based sorting algorithm. 
// It repeatedly steps through the list, compares adjacent elements, 
// and swaps them if they are in the wrong order. The pass through the list 
// is repeated until the array is Sorted.
//
// 🕒 Time Complexity: 
// - Worst Case: O(n^2)  (When Array is Sorted in reverse order)
// - Average Case: O(n^2)
// - Best Case: O(n)     (when array is already Sorted)
// 💾 Space Complexity: O(1) (in-place sorting)
//
// 🧼 Clean and Optimized Implementation: 
// - Uses a 'swapped' flag to stop early if the array becomes sorted mid-pass.
// - No extra space is required.
//
// 🧪 Sample Input/Output:
// Input:  [30, 25, 20, 15, 10]
// Output: [10, 15, 20, 25, 30]
public class BubbleSort{
    
    // this method to use for sorting the array when need for contests
    public static void bubbleSort(int[] array){
        int n= array.length;
        boolean swapped;
        int temp;
        for(int i=0; i< n-1; i++){
            swapped = false;
            for(int j=0; j< n-i-1; j++){
                if (array[j] > array[j+1]) {
                    temp = array[j];
                    array[j] = array[j+1];
                    array[j+1] = temp;
                    swapped = true;
                    
                }
            }
            if(!swapped) break;

        }
    }

    // this method for printing the array
      public static void printArray(int[] array) {
        for (int num : array) {
            System.out.print(num + " ");
        }
        System.out.println();
    }

    public static void main(String[] args){
        int[] array = {30, 25, 20, 15, 10};

        System.out.print("Original array: ");
        printArray(array);


        bubbleSort(array);
        System.out.print("Sorted array:   ");
        printArray(array);
    }
    
    
}